docs/multiple_image_support_design_proposal.md

```diff
@@ -0,0 +1,549 @@
+fragmentation- Feature Name: Multiple Image Support
+- Start Date: 2018-07-06
+- RFC PR:
+
+# Summary
+Firmware for ARMv8-M MCUs are often released in two separate binary images:
+  - secure code
+  - non-secure code
+
+Currently MCUBoot can only handle single firmware image. MCUBoot could be
+extended to handle the secure and non-secure binaries in one single blob and do
+the upgrade of the images in one round.
+This single blob approach has some drawbacks:
+  - It always needs to update both images even if one of them has not changed.
+  - In multi vendor scenario it requires to share the firmware images among the
+    parties. Because the concatenated firmware must be signed by that vendor,
+    who has access to the private key.
+  - If the flash memory of the MCU is not contiguous then it just does not work.
+  - In most of the cases there is padding between the images which is always
+    copied during swap or overwrite, which is an overhead. The smaller the
+    free space is between the images the less the overhead is.
+
+In some use cases it can be desired to update them independently:
+  - Only one of the binaries have changed (i.e.: bug fix). Downloading only the
+    necessary binary uses less power.
+  - The two binaries might be owned by different vendors. They can be in charge
+    to decide when and what to update.
+  - Address the non contiguous memory scenario.
+  - Reduce overhead during update, only swapping necessary part of the flash.
+
+Some difficulties, questions that come up with independent update:
+  - API compliance must be ensured between images to avoid interoperability
+    issues.
+  - Dependency between images must be described, which covers the API alignment
+    issue.
+  - Ensure atomic update when both images must be updated in one go, and
+    firmware cannot be started before updating of both images has properly
+    finished.
+  - In case one of the images needs to be reverted, it might be necessary to
+    revert the other one as well due to API incompatibilities.
+  - Flash will be more fragmented, as two memory slots are needed for each
+    image.
+  - Whether to use shared scratch area between images or use dedicated scratch
+    area per image?
+  - Whether to support image revert at all?
```

Comment:
Using what is called "overwrite-only" mode is probably much simpler, it doesn't require `scratch`, `swap status`, etc. It basically just overwrites slot0 from slot1 (starting from zero on interruption).

Reply:
But still if MCUBoot built with overwrite-only mode the scratch area is reserved just not touched at all, because it is coming from the Zephyr memory config.